pkg/alibabacloud/api/api.go

```diff
@@ -424,6 +430,120 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context) ([]ecs.NetworkIn
 	return result, nil
 }
 
+// describeNetworkInterfacesFromInstances lists all ENIs matching filtered ECS instances.
+// Due to a limitation in the DescribeInstances API, we can only retrieve up to 1,000 instances
+// when filtering by tags directly. To overcome this limitation, an alternative approach is
+// implemented in 3 steps:
+// 1. Filter out matching instance ids with ListTagResources
+// 2. Split instance ids into batches of 100 and send parallel DescribeInstances requests
+// 3. Split eni ids from the instances and send parallel DescribeNetworkInterfaces requests
+// https://www.alibabacloud.com/help/en/ecs/developer-reference/api-ecs-2014-05-26-listtagresources
+// https://www.alibabacloud.com/help/en/ecs/developer-reference/api-ecs-2014-05-26-describeinstances
+// https://www.alibabacloud.com/help/en/ecs/developer-reference/api-ecs-2014-05-26-describenetworkinterfaces
+func (c *Client) describeNetworkInterfacesFromInstances(ctx context.Context) ([]ecs.NetworkInterfaceSet, error) {
+	var result []ecs.NetworkInterfaceSet
+
+	// Get filtered instance IDs
+	tagResouces, err := c.EcsListTagResources(ctx, c.instancesFilters)
+	if err != nil {
+		return nil, err
+	}
+	instanceIds := make([]string, 0, len(tagResouces))
+	for _, t := range tagResouces {
+		instanceIds = append(instanceIds, t.ResourceId)
+	}
+	// The response of ListTagResources can have duplicate instanceId
+	slices.Unique(instanceIds)
+
+	if len(instanceIds) == 0 {
+		return result, nil
+	}
+
+	// DescribeInstances and retrieve the ENI id list. DescribeInstances accepts 100 instance IDs
+	// at most, so split instanceIds into batches of 100 and send parallel requests for performance.
+	// Return error if any request fails.
+	g := new(errgroup.Group)
+	respChan := make(chan *ecs.DescribeInstancesResponse, (len(instanceIds)/100)+1)
+
+	for i := 0; i < len(instanceIds); i += 100 {
+		idx := i
+		endIdx := min(idx+100, len(instanceIds))
+		quotedIds := make([]string, endIdx-idx)
+		for i := idx; i < endIdx; i++ {
+			quotedIds[i-idx] = fmt.Sprintf(`"%s"`, instanceIds[i])
+		}
+
+		g.Go(func() error {
+			req := ecs.CreateDescribeInstancesRequest()
+			// format: ["xxx","xxx","xxx"]
+			req.InstanceIds = fmt.Sprintf("[%s]", strings.Join(quotedIds, ","))
+			req.PageSize = requests.NewInteger(100)
+			c.limiter.Limit(ctx, "DescribeInstances")
```

Comment:
You might want to add constants for all the limiters?

Reply:
Sounds good! Looks like all the limiter names for alibabacloud/aws/azure are currently string literals, what about adding constants for them in a follow-up PR?

Reply:
@doniacld do you consider this conversation resolved? It is blocking the merging of the PR

Reply:
@jaffcheng yes, please address this through a follow-up PR :).

I believe Donia is out, so let's not block this PR for too long on it.